src/Core/PluginRegistry.php

```diff
@@ -0,0 +1,172 @@
+     * @throws \Exception
+     */
+    protected function _throwMissingClassError($class, $plugin)
+    {
+        new RuntimeException(sprintf(
+            'Plugin class `%s` not found.',
+            $class
+        ));
+    }
+
+    /**
+     * Create an instance of a given classname.
+     * This method should construct and do any other initialization logic
+     * required.
+     *
+     * @param string $class The class to build.
+     * @param string $alias The alias of the object.
+     * @param array $config The Configuration settings for construction
+     * @return mixed
+     */
+    protected function _create($class, $alias, $config)
+    {
+        $instance = new $class($config);
+
+        if ((!$instance instanceof HttpApplicationInterface
+            && !$instance instanceof ConsoleApplicationInterface)
+            || !$instance instanceof PluginInterface) {
+            throw new RuntimeException(sprintf(
+                '`%s` is not a valid plugin object. It\'s not implementing `%s` or `%s`',
+                get_class($instance)),
+                HttpApplicationInterface::class,
+                ConsoleApplicationInterface::class
+            );
+        }
+
+        return $instance;
+    }
+
+    /**
+     * Load all the application configuration and bootstrap logic.
+     * Override this method to add additional bootstrap logic for your application.
+     *
+     * @return void
+     */
+    public function bootstrap()
+    {
+        foreach ($this as $plugin) {
+            if ($plugin->isBootstrapEnabled()) {
+                $plugin->bootstrap();
+            }
+        }
+    }
+
+    /**
+     * Define the console commands for an application.
+     *
+     * @param \Cake\Console\CommandCollection $commands The CommandCollection to add commands into.
+     * @return \Cake\Console\CommandCollection The updated collection.
+     */
+    public function console($commands)
+    {
+        foreach ($this as $plugin) {
+            if ($plugin instanceof ConsoleApplicationInterface) {
+                $commands = $plugin->console($commands);
+            }
+        }
+
+        return $commands;
+    }
+
+    /**
+     * Define the routes for an application.
+     * Use the provided RouteBuilder to define an application's routing.
+     *
+     * @param \Cake\Routing\RouteBuilder $routes A route builder to add routes into.
+     * @return void
+     */
+    public function routes($routes)
+    {
+        foreach ($this as $plugin) {
+            if ($plugin->isRouteLoadingEnabled()) {
+                $plugin->routes($routes);
+            }
+        }
+    }
+
+    /**
+     * Define the HTTP middleware layers for an application.
+     *
+     * @param \Cake\Http\MiddlewareQueue $middleware The middleware queue to set in your App Class
+     * @return \Cake\Http\MiddlewareQueue
+     */
+    public function middleware($middleware)
+    {
+        foreach ($this as $plugin) {
+            $middleware->add($plugin);
```

Comment:
Is the thinking here that plugins will *be* middleware? Or that plugins can *add* middleware. I think we want plugins to *add* middleware, which would look like

```php
foreach ($this as $plugin) {
  if ($plugin instanceof HttpApplicationInterface) {
    $middleware = $plugin->middleware($middleware);
  }
}
return $middleware
```